coffee_maker_navigator_2/lib/di/di/dependency_container_manager.dart

```diff
@@ -0,0 +1,137 @@
+
+  /// A map that registers container factories by their type.
+  ///
+  /// This map holds factory functions that are responsible for creating instances
+  /// of dependency containers. Each entry in the map associates a type [T] with a
+  /// factory function that produces a [DependencyContainer] of that type.
+  ///
+  /// The key is the type of the dependency container, and the value is a function
+  /// that takes an existing [AppLevelDependencies] and returns a new instance of a
+  /// [DependencyContainer] for the specified type [T].
+  final _containerFactories =
+      <Type, DependencyContainer Function(AppLevelDependencyContainer)>{};
+
+  // A map that tracks the subscribers for each container type.
+  final _containerSubscribers = <Type, Set<Object>>{};
+
+  // A map that holds the currently active containers by their type.
+  // An active container is one that is currently in use, meaning it has at least one subscriber.
+  final _activeContainers = <Type, DependencyContainer>{};
+
+  /// Returns the single instance of [DependencyContainerManager].
+  static DependencyContainerManager get instance => _instance;
+
+  /// Factory constructor to return the singleton instance.
+  factory DependencyContainerManager() => _instance;
+
+  /// Internal constructor for singleton implementation.
+  DependencyContainerManager._internal();
+
+  /// Initializes the app-level dependencies.
+  Future<void> init() async {
+    await _appLevelDependencies.init();
+    registerContainerFactory<AppLevelDependencyContainer>(
+        (_) => _appLevelDependencies);
+    subscribeToContainer<AppLevelDependencyContainer>(this);
+  }
+
+  /// Registers a dependency container factory for a specific type [T].
+  ///
+  /// This method allows you to register a factory function that will be responsible
+  /// for creating instances of the dependency container for the specified dependency container
+  /// type [T].
+  ///
+  /// [factory]: A function that takes an existing [DependencyContainer] and returns
+  /// a new instance of a [DependencyContainer] for the specified type [T].
+  void registerContainerFactory<T>(
+      DependencyContainer Function(DependencyContainer) factory) {
+    _containerFactories[T] = factory;
+  }
+
+  /// Adds a subscriber for the container of type [C] and manages the container's lifecycle.
+  ///
+  /// This method registers a subscriber for the specified container type [C] and
+  /// manages the lifecycle of the container based on the presence of subscribers.
+  ///
+  /// [subscriber]: An object that subscribes to the container of type [C]. The presence
+  /// of subscribers influences the creation and destruction of the container.
+  void subscribeToContainer<C>(Object subscriber) {
+    final currentSubscribers = _containerSubscribers[C] ??= {};
+    // Add the subscriber to the set of subscribers for the container type C if not already present.
+    if (currentSubscribers.add(subscriber)) {
```

Comment:
in case set is not created yet, it wont be added to _containerSubscribers

Reply:
or if we want to avoid it we can add {} during the registration for the type

Reply:
How come? Check the above line: final currentSubscribers = _containerSubscribers[C] ??= {};

Reply:
a6a0840a2ccf2ec48977cb9b990e5693bc2e9279